### PR TITLE
Add new version 5.0.0 of PyAMG

### PR DESCRIPTION
**Pull request by benc303**
This adds version 5.0.0 of PyAMG. To build it, a newer version of setuptools_scm is needed, so this adds that too.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-setuptools-scm/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-setuptools-scm/package.py:1]: tomli dep needs to be updated, it's no longer needed in Python 3.11

File: var/spack/repos/builtin/packages/py-pyamg/package.py [hunk 1]
```diff
@@ -25,5 +26,7 @@ class PyPyamg(PythonPackage):
     depends_on("py-numpy@1.7:", type=("build", "run"))
     depends_on("py-scipy@0.12:", type=("build", "run"))
     depends_on("py-setuptools@42:", type="build", when="@4.2.0:")
-    depends_on("py-setuptools-scm@5:+toml", type="build", when="@4.2.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyamg/package.py:28]: Is 7.1.0 also needed for 4.2.3 or only for 5.0.0?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-setuptools-scm/package.py
Comment (benc303) [reply var/spack/repos/builtin/packages/py-setuptools-scm/package.py:1]: Confirmed. Removing `+toml` from the dep and then running `spack install py-pyamg^py-setuptools-scm~toml` (on Mac/M1) succeeds. Note that this uses Python 3.10.12.

File: var/spack/repos/builtin/packages/py-pyamg/package.py (hunk 1 above)
Comment (benc303) [reply var/spack/repos/builtin/packages/py-pyamg/package.py:28]: `spack install py-pyamg@4.2^py-setuptools-scm@7.0.5` succeeds. So does `spack install py-pyamg@4.2^py-setuptools-scm@7.0.5~toml`. So apparently, the `+toml` variant isn't required, and 7.1.0 is only required for PyAMG 5.0.0.

File: var/spack/repos/builtin/packages/py-pyamg/package.py [hunk 2]
```diff
@@ -25,5 +26,8 @@ class PyPyamg(PythonPackage):
...
+    # Later version required due to
+    # https://github.com/pypa/setuptools_scm/issues/758
+    depends_on("py-setuptools-scm", type="build", when="@4.2.0:")
+    depends_on("py-setuptools-scm@7.1.0", type="build", when="@5.0.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyamg/package.py]:
```suggestion
    depends_on("py-setuptools-scm@5:+toml", type="build", when="@4.2.0:")
    depends_on("py-setuptools-scm@7.1:+toml", type="build", when="@5.0.0:")
```
Let's stick to the documented requirements as best as possible.

File: var/spack/repos/builtin/packages/py-setuptools-scm/package.py [hunk 3]
```diff
@@ -12,6 +12,7 @@ class PySetuptoolsScm(PythonPackage):
     homepage = "https://github.com/pypa/setuptools_scm"
     pypi = "setuptools_scm/setuptools_scm-4.1.2.tar.gz"
 
+    version("7.1.0", sha256="6c508345a771aad7d56ebff0e70628bf2b0ec7573762be9960214730de278f27")
     version("7.0.5", sha256="031e13af771d6f892b941adb6ea04545bbf91ebc5ce68c78aaf3fff6e1fb4844")
     version("7.0.3", sha256="cf8ab8e235bed840cd4559b658af0d8e8a70896a191bbc510ee914ec5325332d")
     version("6.3.2", sha256="a49aa8081eeb3514eb9728fa5040f2eaa962d6c6f4ec9c32f6c1fba88f88a0f2")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-setuptools-scm/package.py:18]: tomli dep needs to be updated:
```python
depends_on("py-tomli@1:", when="@7.1: ^python@:3.10", type=("build", "run"))
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-setuptools-scm/package.py (hunk 3 above)
Comment (benc303) [reply var/spack/repos/builtin/packages/py-setuptools-scm/package.py:18]: There's already
```
depends_on("py-tomli@1:", when="@7:", type=("build", "run"))
```
Should that be
```
depends_on("py-tomli@1:", when="@7.0", type=("build", "run"))
```
?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-setuptools-scm/package.py:18]: Yes

File: var/spack/repos/builtin/packages/py-setuptools-scm/package.py [hunk 4]
```diff
@@ -37,6 +38,7 @@ class PySetuptoolsScm(PythonPackage):
     depends_on("py-setuptools@34.4:", type=("build", "run"))
     depends_on("py-toml", when="+toml @:6.1.0", type=("build", "run"))
     depends_on("py-tomli@1:", when="+toml @6.1.0:", type=("build", "run"))
-    depends_on("py-tomli@1:", when="@7:", type=("build", "run"))
+    depends_on("py-tomli@1:", when="@7.0:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-setuptools-scm/package.py]:
```suggestion
    depends_on("py-tomli@1:", when="@7.0", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-pyamg/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-pyamg/package.py]: Still waiting on this change

Comment (benc303) [reply var/spack/repos/builtin/packages/py-pyamg/package.py]: Done.

File: var/spack/repos/builtin/packages/py-pyamg/package.py [hunk 5]
```diff
@@ -25,5 +26,8 @@ class PyPyamg(PythonPackage):
     depends_on("py-numpy@1.7:", type=("build", "run"))
     depends_on("py-scipy@0.12:", type=("build", "run"))
     depends_on("py-setuptools@42:", type="build", when="@4.2.0:")
+    # Later version required due to
+    # https://github.com/pypa/setuptools_scm/issues/758
     depends_on("py-setuptools-scm@5:+toml", type="build", when="@4.2.0:")
+    depends_on("py-setuptools-scm@7.1:+toml", type="build", when="@5.0.0:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyamg/package.py]: Let's swap these two lines so we know the comment is for the second line

Comment (benc303) [reply var/spack/repos/builtin/packages/py-pyamg/package.py]: Done.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**